### ISSUE TITLE
KJ_SYSCALL(close(...)) results in re-trying close on EINTR, which is not correct.

### ISSUE THREAD

**Issue by zenhack** [sandstorm-io/sandstorm#2928]
This is all in `supervisor.c++`. Granted, it shouldn't cause the problem described below in a single-threaded program, which I *think* the supervisor is?

Per `close(2)`:

> Retrying  the close() after a failure return is the wrong thing to do, since this may cause a reused file  descriptor from another thread to be closed.  This can occur because the Linux kernel always releases the file descriptor early in the close operation, freeing it for reuse; the steps that may return an error, such as flushing data to the filesystem or device, occur only later in the close operation.

It goes on to discuss the fact that correct behaviour is different on different platforms, and POSIX doesn't give developers a useful out.

Again, shouldn't be an issue in a single-threaded program. But, if EINTR, *does* get returned, my understanding is that (assuming the file descriptor doesn't get re-allocated) the retry would get EBADF, which would raise an exception.

Using KJ_SYSCALL for close is probably the wrong thing.

Note that this happens in a few places in the capnproto compiler as well, but in that case it needs to run on systems other than Linux, so it's a bit murkier.

**Comment by kentonv** [comment]
Heh, I remember about a decade ago, when the man page didn't document this at all, spending a day scouring mailing lists to find out exactly what happens when close() fails with EINTR. Note that e.g. this code predates Sandstorm entirely: https://github.com/sandstorm-io/capnproto/blob/master/c++/src/kj/io.c++#L286

But KJ_SYSCALL is so easy to write and EINTR never really happens in practice so I get lazy sometimes. :/

The right thing to do would probably be to use AutoCloseFd more systematically, which would be easier if all of the syscalls that create FDs had proper KJ wrappers. I have been slowly writing KJ wrappers for more and more OS functionality...

Amusingly, the POSIX standard now declares that Linux's behavior is wrong. Oops.

**Comment by zenhack** [comment]
For whatever reason I was thinking about this again recently. We could make KJ_SYSCALL usable as an expression. This would make the call sites more ergonomic, as you could just do e.g:

```c++
AutoCloseFd fd = KJ_SYSCALL(open(...));
```

We could maybe also add a wrapper `KJ_SYSCALL_FD` that is just `AutoCloseFd(KJ_SYSCALL(...))`, would make this even nicer, since you could use `auto` and such.

I took a stab at an implementation, this passes all the tests:

```c++
diff --git a/c++/src/kj/debug.h b/c++/src/kj/debug.h
index 77b8fe2d..5ffc0490 100644
--- a/c++/src/kj/debug.h
+++ b/c++/src/kj/debug.h
@@ -146,9 +146,18 @@ namespace kj {
                                nullptr, "" #__VA_ARGS__, __VA_ARGS__);; f.fatal())
 
 #define KJ_SYSCALL(call, ...) \
-  if (auto _kjSyscallResult = ::kj::_::Debug::syscall([&](){return (call);}, false)) {} else \
-    for (::kj::_::Debug::Fault f(__FILE__, __LINE__, \
-             _kjSyscallResult.getErrorNumber(), #call, "" #__VA_ARGS__, __VA_ARGS__);; f.fatal())
+  ::kj::_::syscallRetryEINTR ( \
+    [&]() { return (call); }, \
+    [&](int errorNumber) { \
+      ::kj::_::Debug::Fault f( \
+          __FILE__, \
+          __LINE__, \
+          errorNumber, \
+          #call, \
+          "" #__VA_ARGS__, __VA_ARGS__); \
+      f.fatal(); \
+    } \
+  )
 
 #define KJ_NONBLOCKING_SYSCALL(call, ...) \
   if (auto _kjSyscallResult = ::kj::_::Debug::syscall([&](){return (call);}, true)) {} else \
@@ -485,6 +494,20 @@ private:
   // Get the error code of the last error (e.g. from errno).  Returns -1 on EINTR.
 };
 
+template <class Call, class OnError>
+static inline int syscallRetryEINTR(Call&& call, OnError onError) {
+  int ret;
+  int errorNumber;
+  do {
+    ret = call();
+    errorNumber = Debug::syscall(kj::mv(call), false);
+  } while(ret < 0 && errorNumber == -1);
+  if(ret < 0) {
+    onError(errorNumber);
+  }
+  return ret;
+}
+
 template <typename... Params>
 void Debug::log(const char* file, int line, LogSeverity severity, const char* macroArgs,
                 Params&&... params) {

```

If the approach sounds reasonable to you I'll submit a PR.

**Comment by kentonv** [comment]
I'd love for `KJ_SYSCALL()` to be an expression, but doesn't your change break syscalls that have fallback blocks? Like:

    KJ_SYSCALL(foo()) {
      // When exceptions are disabled we should just return
      // early rather than crash.
      return;
    }

IIRC I couldn't figure out any way to make KJ_SYSCALL be an expression while still supporting fallback blocks.

**Comment by zenhack** [comment]
Ah, I didn't realize you could do that (I was confused about what the for loop was for...). Yeah, I don't see an obvious way to support that syntax and use as an expression with the same macro.

**Comment by kentonv** [comment]
Yes. Though you said the tests pass, which I guess implies that no code anywhere in KJ or Sandstorm actually uses a fallback block for `KJ_SYSCALL`? I'm a bit surprised by that (there are definitely many asserts that use fallback blocks), but if that's true then it suggests maybe that functionality isn't important, whereas returning a value would be pretty nice.

I also increasingly wonder whether I could stop trying to support disabling exceptions. It's pretty cumbersome trying to make sure to use fallback blocks all over the place. Plus GCC and Clang still both have bugs where if you try to return a non-trivial value from a fallback block, you'll leak memory in the case that exceptions _are_ enabled, because the compiler forgets to invoke the return value's destructor when the exception is thrown.

OTOH fallback blocks are important even when exceptions are enabled, in the case that an assert fires inside a destructor while another unwind is already taking place. In that case we skip throwing the exception since it will probably terminate the program. But the handling of exception-while-unwinding is the single worst feature of C++, to the point where I am seriously considering trying to monkey-patch the exception-handling logic to change that behavior under KJ... in which case maybe fallback blocks aren't needed for that case anymore.

**Comment by zenhack** [comment]
Yep, everything compiles, all the tests pass, light manual testing of sandstorm goes well.

*Sandstorm* certainly doesn't care about disabling exceptions. I'll leave it to you how much you care about avoiding breaking compatibility for other projects' sake.
